Tools/AP_Bootloader/network.cpp

```diff
@@ -0,0 +1,631 @@
+        sock->send(str, strlen(str));
+    } else {
+        status_printf("Flash done: ERR:%u", unsigned(ok));
+    }
+}
+
+/*
+  handle an incoming HTTP request
+ */
+void BL_Network::handle_request(SocketAPM *sock)
+{
+    /*
+      read HTTP headers
+     */
+    char *headers = read_headers(sock);
+
+    const char *header = "HTTP/1.1 200 OK\r\n"
+        "Content-Type: text/html\r\n"
+        "Connection: close\r\n"
+        "\r\n";
+    sock->send(header, strlen(header));
+
+    if (strncmp(headers, "POST / ", 7) == 0) {
+        const char *clen = "\r\nContent-Length:";
+        const char *p = strstr(headers, clen);
+        if (p != nullptr) {
+            p += strlen(clen);
+            const uint32_t content_length = atoi(p);
+            handle_post(sock, content_length);
+            delete headers;
+            delete sock;
+            return;
+        }
+    }
+
+    /*
+      check for async status
+     */
+    const char *get_status = "GET /bootloader_status.html";
+    if (strncmp(headers, get_status, strlen(get_status)) == 0) {
+        {
+            WITH_SEMAPHORE(status_mtx);
+            sock->send(bl_status, strlen(bl_status));
+        }
+        delete headers;
+        delete sock;
+        return;
+    }
+
+    const char *get_reboot = "GET /REBOOT";
+    if (strncmp(headers, get_reboot, strlen(get_reboot)) == 0) {
+        need_reboot = true;
+    }
+    
+    uint32_t size = 0;
+    const auto *msg = AP_ROMFS::find_decompress("index.html", size);
+    if (need_reboot) {
+        const char *str = "<html><head><meta http-equiv=\"refresh\" content=\"2; url=/\"></head></html>";
+        sock->send(str, strlen(str));
+    } else {
+        char *msg2 = substitute_vars((const char *)msg, size);
+        sock->send(msg2, strlen(msg2));
+        delete msg2;
+    }
+    delete headers;
+    delete sock;
+    AP_ROMFS::free(msg);
+}
+
+struct req_context {
+    BL_Network *driver;
+    SocketAPM *sock;
+};
+
+void BL_Network::net_request_trampoline(void *ctx)
+{
+    auto *req = (req_context *)ctx;
+    req->driver->handle_request(req->sock);
+    delete req;
+}
+
+/*
+  web server thread
+ */
+void BL_Network::web_server(void)
+{
+    auto *listen_socket = new SocketAPM(0);
+    listen_socket->bind("0.0.0.0", 80);
+    listen_socket->listen(20);
+
+    while (true) {
+        auto *sock = listen_socket->accept(20);
+        if (need_reboot) {
+            need_reboot = false;
+            NVIC_SystemReset();
+        }
+        if (need_launch) {
+            need_launch = false;
+            thread_sleep_ms(1000);
+            jump_to_app();
+        }
+        if (sock == nullptr) {
+            continue;
+        }
+        // a neq thread for each connection to allow for AJAX
+        auto *req = new req_context;
+        req->driver = this;
+        req->sock = sock;
+        thread_create_alloc(THD_WORKING_AREA_SIZE(2048),
+                            "net_request",
```

Comment:
This is an odd name for a web server thread

Reply:
i've renamed to web_request